In [2]:
# Import necessary libraries
import pandas as pd
import sqlite3

# Upload the dataset to Colab
from google.colab import files
uploaded = files.upload()

# Read the dataset into a pandas DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database and write the DataFrame to it
conn = sqlite3.connect('hotel_reservations.db')
df.to_sql('reservations', conn, if_exists='replace', index=False)

# Create a cursor object to execute SQL queries
cur = conn.cursor()

Saving Hotel Reservation Dataset.csv to Hotel Reservation Dataset.csv


1. What is the total number of reservations in the dataset?

In [15]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT SUM(no_of_adults) AS total_adults, SUM(no_of_children) AS total_children FROM hotel_reservations;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

(1316, 69)


Hence, total 1316 + 69 = 1385 hotel reservations

**1 To count the confirmed and cancelled reservations of adults and children :

In [16]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT booking_status, SUM(no_of_adults) AS total_adults, SUM(no_of_children) AS total_children FROM hotel_reservations GROUP BY booking_status;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

('Canceled', 402, 31)
('Not_Canceled', 914, 38)


Hence, total cancelled reservations = 402 + 31 = 433 and total confirmed reservations = 914 + 38 = 952

2. Which meal plan is the most popular among guests?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT type_of_meal_plan, COUNT(Booking_ID) AS reservation_count FROM hotel_reservations GROUP BY type_of_meal_plan ORDER BY reservation_count DESC LIMIT 1;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

('Meal Plan 1', 527)


3. What is the average price per room for reservations involving children?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT round(AVG(avg_price_per_room)) AS avg_price FROM hotel_reservations WHERE no_of_children > 0;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

(145.0,)


4. How many reservations were made for the year 2018 ?


In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT COUNT(Booking_ID) AS Total_Reservation_2018 FROM hotel_reservations WHERE arrival_date like '%2018%';")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

(577,)


5. What is the most commonly booked room type?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT room_type_reserved, COUNT(Booking_ID) AS reservation_count FROM hotel_reservations GROUP BY room_type_reserved ORDER BY reservation_count DESC LIMIT 1;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

('Room_Type 1', 534)


6. How many reservations fall on a weekend ?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT COUNT(no_of_weekend_nights) AS weekend_reservations FROM hotel_reservations WHERE no_of_weekend_nights > 0;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

(383,)


7. What is the highest and lowest lead time for reservations?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT MAX(lead_time) AS max_lead_time, MIN(lead_time) AS min_lead_time FROM hotel_reservations;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

(443, 0)


8. What is the most common market segment type for reservations?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT market_segment_type, COUNT(Booking_ID) AS reservation_count FROM hotel_reservations GROUP BY market_segment_type ORDER BY reservation_count DESC LIMIT 1;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

('Online', 518)


9. How many reservations have a booking status of "Confirmed"?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT COUNT(Booking_ID) AS confirmed_reservations FROM hotel_reservations WHERE booking_status = 'Not_Canceled';")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

(493,)


10. What is the total number of adults and children across all reservations?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT SUM(no_of_adults) AS total_adults, SUM(no_of_children) AS total_children FROM hotel_reservations;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

(1316, 69)


11. What is the average number of weekend nights for reservations involving children?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT AVG(no_of_weekend_nights) AS avg_weekend_nights FROM hotel_reservations WHERE no_of_children > 0;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

(1.0,)


12. How many reservations were made in each month of the year?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT arrival_date as month,COUNT(Booking_ID) as reservation_per_month FROM hotel_reservations group by arrival_date")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

('01-03-2018', 1)
('01-04-2018', 5)
('01-05-2018', 2)
('01-06-2018', 1)
('01-07-2017', 2)
('01-07-2018', 2)
('01-08-2018', 3)
('01-09-2018', 1)
('01-10-2017', 1)
('01-10-2018', 2)
('01-11-2018', 4)
('02-01-2018', 3)
('02-04-2018', 1)
('02-05-2018', 3)
('02-07-2018', 1)
('02-09-2017', 1)
('02-09-2018', 2)
('02-10-2017', 4)
('02-10-2018', 2)
('02-11-2017', 2)
('02-11-2018', 1)
('02-12-2018', 4)
('03-02-2018', 1)
('03-03-2018', 1)
('03-04-2018', 2)
('03-05-2018', 1)
('03-06-2018', 3)
('03-08-2018', 4)
('03-09-2017', 1)
('03-09-2018', 3)
('03-10-2018', 1)
('03-11-2018', 3)
('03-12-2018', 2)
('04-02-2018', 1)
('04-03-2018', 2)
('04-04-2018', 4)
('04-05-2018', 2)
('04-07-2018', 1)
('04-08-2018', 2)
('04-09-2017', 1)
('04-09-2018', 1)
('04-10-2017', 1)
('04-10-2018', 1)
('04-11-2018', 2)
('05-02-2018', 1)
('05-03-2018', 2)
('05-04-2018', 2)
('05-05-2018', 3)
('05-06-2018', 2)
('05-07-2018', 1)
('05-09-2017', 2)
('05-09-2018', 2)
('05-10-2017', 1)
('05-10-2018', 1)
('05-11-2018', 2)
('05-12-20

13. What is the average number of nights (both weekend and weekday) spent by guests for each room type?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT room_type_reserved, round(AVG(no_of_weekend_nights + no_of_week_nights)) AS avg_total_nights FROM hotel_reservations GROUP BY room_type_reserved;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

('Room_Type 1', 3.0)
('Room_Type 2', 3.0)
('Room_Type 4', 4.0)
('Room_Type 5', 3.0)
('Room_Type 6', 4.0)
('Room_Type 7', 3.0)


14. For reservations involving children, what is the most common room type, and what is the average price for that room type?

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("WITH child_reservations AS (SELECT room_type_reserved, avg_price_per_room FROM hotel_reservations WHERE no_of_children > 0) SELECT room_type_reserved, round(AVG(avg_price_per_room)) AS avg_price FROM child_reservations GROUP BY room_type_reserved ORDER BY COUNT(*) DESC LIMIT 1;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

('Room_Type 1', 123.0)


15. Find the market segment type that generates the highest average price per room.

In [ ]:
# Read the CSV file into a DataFrame
df = pd.read_csv('Hotel Reservation Dataset.csv')

# Create a SQLite database in memory
con = sqlite3.connect(':memory:')

# Write the DataFrame to the database
df.to_sql('hotel_reservations', con, if_exists='replace', index=False)

# Execute SQL queries on the database
cur = con.cursor()
cur.execute("SELECT market_segment_type, round(AVG(avg_price_per_room)) AS avg_price FROM hotel_reservations GROUP BY market_segment_type ORDER BY avg_price DESC LIMIT 1;")
rows = cur.fetchall()

# Print the results
for row in rows:
  print(row)

# Close the connection to the database
con.close()

('Online', 112.0)
